src/System.Text.Primitives/System/Text/Encoders/Utf8Util.Validation.cs

```diff
@@ -0,0 +1,728 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+
+//
+// This file contains workhorse methods for performing validation of UTF-8 byte sequences.
+//
+
+namespace System.Buffers.Text
+{
+    internal static partial class Utf8Util
+    {
+        // This method will consume as many ASCII bytes as it can using fast vectorized processing, returning the number of
+        // consumed (ASCII) bytes. It's possible that the method exits early, perhaps because there is some non-ASCII byte
+        // later in the sequence or because we're running out of input to search. The intent is that the caller *skips over*
+        // the number of bytes returned by this method, then it continues data processing from the next byte.
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static unsafe int ConsumeAsciiBytesVectorized(ref byte buffer, int length)
+        {
+            // Only allow vectorization if vector operations are hardware-accelerated
+            // and vectors are at least 128 bits in length.
+            if (Vector.IsHardwareAccelerated && (Vector<byte>.Count >= 2 * sizeof(ulong)))
+            {
+                // Need to pin buffer so the GC doesn't cause it to be misaligned when we're
+                // reading vectors from it.
```

Comment:
I heard there is no reason to do it. The perf penalty is not significant.
cc: @benaadams

Reply:
I saw that the JITter produced VMOVDQU, so unaligned access should be safe, but I wasn’t sure if this was an implementation detail or not. I should perf test VMOVDQU vs VMOVDQA when the intrinsics come online.

Reply:
The idiv will likely be more significant :) `(int)pbBuffer % Vector<byte>.Count`

Is a perfhit if you cross cache lines (worse if you cross pages); so could go for 64 byte alignment and play it safe for all vector lengths?

```csharp
int referenceMisalignment = (int)pbBuffer & 63;
```

Issue if they go bigger than 512 bytes, but will also be larger issues as most algos are optimized for cache lines of 64 bytes

Reply:
Yup, JIT optimizes the modulus operation since the value is known at codegen time. Sounds like I’ll still need to pin the buffer though since the GC doesn’t guarantee alignment beyond the natural word size, right?